### PR TITLE
feat: assign primary ip to a server on create

### LINKED ISSUE #464: Create Primary IP on Server
##### SUMMARY
I cannot enable a Public IP on an existing server that has it disabled.  I should be able to from the CLI documentation.

https://docs.hetzner.cloud/#primary-ips-create-a-primary-ip

##### ISSUE TYPE
- Feature Idea

##### COMPONENT NAME
`hetzner.hcloud.primary_ip`

##### ADDITIONAL INFORMATION

```yaml
- name: Create a basic IPv4 Primary IP
  hetzner.hcloud.primary_ip:
    name: "{{ server.hcloud_server.name }}"
    assignee_id: "{{ server.hcloud_server.id }}"
    type: ipv4
    auto_delete: true

# API:
- name: Enable public IP
  ansible.builtin.uri:
    url: "https://api.hetzner.cloud/v1/primary_ips"
    method: POST
    body_format: json
    body:
      assignee_id: "{{ server.hcloud_server.id }}"
      assignee_type: "server"
      type: "ipv4"
      name: "{{ server.hcloud_server.name }}"
      auto_delete: true
    status_code: 201
    headers:
      Authorization: "Bearer {{ HETZNER_TOKEN }}"
      Content-Type: "application/json"

```

### PR DESCRIPTION
**Pull request by jjeff07**
##### SUMMARY

Fixes #464 

##### ISSUE TYPE
- Feature Pull Request

##### COMPONENT NAME
primary_ip

### CODE REVIEW

File: plugins/modules/primary_ip.py [hunk 1]
```diff
@@ -163,13 +204,20 @@ def _get_primary_ip(self):
...
-                "datacenter": self.client.datacenters.get_by_name(self.module.params.get("datacenter")),
+                "auto_delete": self.module.params.get("auto_delete"),
+                "datacenter": None,
             }
+            if self.module.params.get("server") is not None:
+                params["assignee_id"] = self._client_get_by_name_or_id("servers", self.module.params.get("server")).id
+            elif self.module.params.get("datacenter") is not None:
+                params["datacenter"] = self.client.datacenters.get_by_name(self.module.params.get("datacenter"))
+            else:
+                self.module.fail_json(msg="server or datacenter is required")
```
Comment (jooola) [inline plugins/modules/primary_ip.py:220]: Maybe using `check_required_one_of` from the `ansible.module_utils.common.validation` module is a better idea here. Not sure if this is already in ansible-core 2.14 though.

File: tests/integration/targets/server/tasks/test_primary_ips.yml [hunk 2]
```diff
@@ -57,15 +57,13 @@
     that:
       - result is changed
 
-- name: cleanup server with primary ips
-  hetzner.hcloud.server:
-    name: "{{ hcloud_server_name }}"
-    state: absent
-
```
Comment (jooola) [inline tests/integration/targets/server/tasks/test_primary_ips.yml:64]: Why did you remove this?

File: tests/integration/targets/server/tasks/test_primary_ips.yml [hunk 3]
```diff
@@ -57,15 +57,13 @@
...
 - name: cleanup test create primary ips
   hetzner.hcloud.primary_ip:
     name: "{{ hcloud_server_name }}v4"
     state: absent
+  until: result is not failed
+  retries: 5
+  delay: 2
```
Comment (jooola) [inline tests/integration/targets/server/tasks/test_primary_ips.yml:66]: Why is this required?

File: tests/integration/targets/server/tasks/test_primary_ips.yml [hunk 4]
```diff
@@ -80,3 +78,44 @@
...
+  hetzner.hcloud.primary_ip:
+    name: "{{ hcloud_primary_ip_name }}v4"
+    server: "{{ hcloud_server_name }}"
+    type: ipv4
+    state: present
+- name: verify test create primary ipv4 attached to server
+  assert:
+    that:
+      - result is changed
+- name: test create primary ipv6 attached to server
```
Comment (jooola) [inline tests/integration/targets/server/tasks/test_primary_ips.yml:92]:
```suggestion
      - result is changed
      
- name: test create primary ipv6 attached to server
```

File: tests/integration/targets/server/tasks/test_primary_ips.yml [hunk 5]
```diff
@@ -80,3 +78,44 @@
...
+- name: test create primary ipv4 attached to server
+  hetzner.hcloud.primary_ip:
+    name: "{{ hcloud_primary_ip_name }}v4"
+    server: "{{ hcloud_server_name }}"
+    type: ipv4
+    state: present
+- name: verify test create primary ipv4 attached to server
+  assert:
+    that:
+      - result is changed
```
Comment (jooola) [inline tests/integration/targets/server/tasks/test_primary_ips.yml:91]: Could we test the newly added return values too?

**Review by jooola — CHANGES_REQUESTED** [review]
Awesome, thanks for the PR.

There are a few things to address before we can merge this.

2 changelog fragments are missing for the 2 changes implemented in this PR: https://github.com/ansible-collections/hetzner.hcloud/tree/main/changelogs/fragments

File: plugins/modules/primary_ip.py (hunk 1 above)
Comment (jjeff07) [reply plugins/modules/primary_ip.py:220]: https://docs.ansible.com/ansible/latest/porting_guides/porting_guide_core_2.11.html

File: tests/integration/targets/server/tasks/test_primary_ips.yml (hunk 2 above)
Comment (jjeff07) [reply tests/integration/targets/server/tasks/test_primary_ips.yml:64]: Moved to end of tests to re-use server instead of creating a new one.
https://github.com/ansible-collections/hetzner.hcloud/blob/8534108d791776f22faae307dc3c860d3445ae59/tests/integration/targets/server/tasks/test_primary_ips.yml#L118C1-L121C18

File: tests/integration/targets/server/tasks/test_primary_ips.yml (hunk 3 above)
Comment (jjeff07) [reply tests/integration/targets/server/tasks/test_primary_ips.yml:66]: Not sure all the other tests had it except that one.  I can remove but was just standardizing the removals.
https://github.com/ansible-collections/hetzner.hcloud/blob/694a238a386f6b03013f8e852e22c48f0d40f58d/tests/integration/targets/server/tasks/test_primary_ips.yml#L65C1-L82C11

File: plugins/modules/primary_ip.py [hunk 6]
```diff
@@ -163,13 +204,20 @@ def _get_primary_ip(self):
...
     def _create_primary_ip(self):
-        self.module.fail_on_missing_params(required_params=["type", "datacenter"])
+        self.module.fail_on_missing_params(required_params=["type", "name"])
         try:
             params = {
                 "type": self.module.params.get("type"),
                 "name": self.module.params.get("name"),
-                "datacenter": self.client.datacenters.get_by_name(self.module.params.get("datacenter")),
+                "auto_delete": self.module.params.get("auto_delete"),
+                "datacenter": None,
```
[the remark this answers is not in the document]
Comment (jooola) [reply plugins/modules/primary_ip.py]: This feels dirty, but it is actually a good workaround until we fixed: https://github.com/hetznercloud/hcloud-python/pull/363

Could you add a comment on why we store the None value here?

File: plugins/modules/primary_ip.py (hunk 1 above)
Comment (jjeff07) [reply plugins/modules/primary_ip.py:227]: Could you provide some input on this?  Not really sure how to implement and cannot find any examples.

**Comment by jjeff07** [comment]
@jooola New tests are passing: https://dev.azure.com/ansible/hetzner.hcloud/_build/results?buildId=104488&view=logs&j=2cdf8f08-b819-5dce-9da1-1512c6cd89ab&t=faae668e-70a5-53d6-d16a-f1df6c5fe16b&l=1091

Tests:

- Create IPv4 attached to server
- Check IPv4 assignee_id and auto_delete is false
- Create IPv6 on server set to auto_delete
- Check IPv6 assignee_id and auto_delete set to true
- Remove IPv4
- Remove server
- Check IPv6 was deleted due to `auto_delete`

File: plugins/modules/primary_ip.py [hunk 7]
```diff
@@ -127,6 +154,21 @@
             sample:
                 key: value
                 mylabel: 123
+        assignee_id:
+            description: ID of the resource the Primary IP is assigned to, null if it is not assigned.
+            type: int
+            returned: always
+            sample: 1937415
+        assignee_type:
+            description: Resource type the Primary IP can be assigned to.
```
Comment (jooola) [inline plugins/modules/primary_ip.py:163]:
```suggestion
            description: Resource type the Primary IP is assigned to.
```

File: plugins/modules/primary_ip.py (hunk 1 above)
Comment (jooola) [reply plugins/modules/primary_ip.py:227]: I have never used it either, but you could try to do add this under the existing validation for the create method:
```py
from ansible.module_utils.common.validation import check_required_one_of
from ansible.module_utils.errors import RequiredOneOfError

#...
    def _create_primary_ip(self):
        self.module.fail_on_missing_params(required_params=["type", "name"])
        try:
            check_required_one_of(["server", "datacenter"], self.module.params)
        except TypeError as exc:
            raise RequiredOneOfError(exc) from exc
```

Haven't tested it though. Maybe reraising the TypeError as RequiredOneOfError is not really useful.

**Review by jooola — CHANGES_REQUESTED** [review]

File: plugins/modules/primary_ip.py (hunk 7 above)
Comment (jjeff07) [reply plugins/modules/primary_ip.py:163]: From docs:

File: plugins/modules/primary_ip.py (hunk 1 above)
Comment (jjeff07) [reply plugins/modules/primary_ip.py:227]: Looks like it does not need the try/except, same results.

Without:
`check_required_one_of(["server", "datacenter"], self.module.params)`
```
TASK [test] *******************************************************************
task path: /home/justin/sa-automation/playbooks/test.yml:10
fatal: [localhost]: FAILED! => changed=false
  invocation:
    module_args:
      api_endpoint: https://api.hetzner.cloud/v1
      api_token: VALUE_SPECIFIED_IN_NO_LOG_PARAMETER
      auto_delete: false
      datacenter: null
      delete_protection: null
      id: null
      labels: null
      name: primary-ip-test
      server: null
      state: present
      type: ipv4
  msg: server or datacenter is required
...ignoring
Tuesday 20 February 2024  12:13:33 -0500 (0:00:00.841)       0:00:01.646 ******

TASK [debug] ***********************************************************************************************************
task path: /home/justin/sa-automation/playbooks/test.yml:18
ok: [localhost] =>
  msg:
    changed: false
    failed: true
    msg: server or datacenter is required
```

With try:

```
try:
            check_required_one_of(["server", "datacenter"], self.module.params)
except TypeError as exc:
            raise RequiredOneOfError(exc) from exc
```
```
TASK [test] *******************************************************************
task path: /home/justin/sa-automation/playbooks/test.yml:10
fatal: [localhost]: FAILED! => changed=false
  invocation:
    module_args:
      api_endpoint: https://api.hetzner.cloud/v1
      api_token: VALUE_SPECIFIED_IN_NO_LOG_PARAMETER
      auto_delete: false
      datacenter: null
      delete_protection: null
      id: null
      labels: null
      name: primary-ip-test
      server: null
      state: present
      type: ipv4
  msg: server or datacenter is required
...ignoring
Tuesday 20 February 2024  12:15:08 -0500 (0:00:01.371)       0:00:02.284 ******

TASK [debug] ***********************************************************************************************************
task path: /home/justin/sa-automation/playbooks/test.yml:18
ok: [localhost] =>
  msg:
    changed: false
    failed: true
    msg: server or datacenter is required
```

File: plugins/modules/primary_ip.py (hunk 7 above)
Comment (jooola) [reply plugins/modules/primary_ip.py:163]: I believe the documentation has a few mistakes, I'll discuss this internally with the responsible team.

File: plugins/modules/primary_ip.py (hunk 1 above)
Comment (jjeff07) [reply plugins/modules/primary_ip.py]: Ok that function and most of those validators are dumb.  They check if parameter is present but not set however figured it out.

Comment (jjeff07) [reply plugins/modules/primary_ip.py]: Copied this method:

```python
    def fail_on_missing_params(self, required_params=None):
        if not required_params:
            return
        try:
            check_missing_parameters(self.params, required_params)
        except TypeError as e:
            self.fail_json(msg=to_native(e))
```

Which becomes:
```python
    def _create_primary_ip(self):
        self.module.fail_on_missing_params(required_params=["type", "name"])
        try:
            check_required_one_of([["server", "datacenter"]], {k: v for k, v in self.module.params.items() if v})
        except TypeError as e:
            self.module.fail_json(msg=to_native(e))
```

New test:
```yaml
- name: test no datacenter or server
  hetzner.hcloud.primary_ip:
    name: "{{ hcloud_primary_ip_name }}fail"
    type: ipv4
    state: present
  ignore_errors: true
  register: result
- name: verify test no datacenter or server
  assert:
    that:
      - result is failed
      - 'result.msg == "one of the following is required: server, datacenter"'
```

```
08:47 TASK [server : test no datacenter or server] ***********************************
08:48 fatal: [testhost]: FAILED! => {"changed": false, "msg": "one of the following is required: server, datacenter"}
08:48 ...ignoring
08:48 
08:48 TASK [server : verify test no datacenter or server] ****************************
08:48 ok: [testhost] => {
08:48     "changed": false,
08:48     "msg": "All assertions passed"
08:48 }
```

File: tests/integration/targets/server/tasks/test_primary_ips.yml [hunk 8]
```diff
@@ -51,32 +51,140 @@
...
+
+- name: cleanup server with primary ips
+  hetzner.hcloud.server:
+    name: "{{ hcloud_server_name }}"
+    state: absent
+  register: server
+  ignore_errors: true
+- name: cleanup server with primary ips - retry due to many timeouts found during testing
```
Comment (jjeff07) [inline tests/integration/targets/server/tasks/test_primary_ips.yml:154]: There seems to be a bug that needs investigating.  I am facing many timeouts when I delete a server with a Primary IP set to `auto_delete`.  This is seen both in the pipelines and in my local environment.

In order for the tests to pass I have ignored the error, added a retry, and asserting the server does not exist.

Comment (jjeff07) [reply tests/integration/targets/server/tasks/test_primary_ips.yml:154]: See jobs:
- [20240220.7](https://dev.azure.com/ansible/hetzner.hcloud/_build/results?buildId=104619)
- [20240220.6](https://dev.azure.com/ansible/hetzner.hcloud/_build/results?buildId=104613)

File: plugins/modules/primary_ip.py (hunk 7 above)
Comment (jooola) [reply plugins/modules/primary_ip.py:163]: You are correct, I had a bad understanding of the description of this field.

File: plugins/modules/primary_ip.py [hunk 9]
```diff
@@ -163,13 +212,22 @@ def _get_primary_ip(self):
             self.fail_json_hcloud(exception)
 
     def _create_primary_ip(self):
-        self.module.fail_on_missing_params(required_params=["type", "datacenter"])
+        self.module.fail_on_missing_params(required_params=["type", "name"])
+        try:
+            check_required_one_of([["server", "datacenter"]], {k: v for k, v in self.module.params.items() if v})
+        except TypeError as e:
+            self.module.fail_json(msg=to_native(e))
```
Comment (jooola) [inline plugins/modules/primary_ip.py:219]: I opened https://github.com/ansible/ansible/issues/82724 to see if we can improve this.

Comment (jooola) [reply plugins/modules/primary_ip.py:219]: Let's implement our own and move this code to a helper method in the parent class, as I doubt ansible-core will ship the fix in the next weeks. @jjeff07 still up to continue this PR?

**Comment by jooola** [comment]
@jjeff07 Thanks for the initial work!
